pullman/helpers.go

```diff
@@ -57,3 +62,98 @@ func HashStrings(strings ...string) string {
 
 	return fmt.Sprintf("%#x", h.Sum64())
 }
+
+// Extract a zip file given a file path. Stream reading zips is unsupported.
+func ExtractZip(zipFilePath string, dest string) error {
+	zipReader, err := zip.OpenReader(zipFilePath)
+	if err != nil {
+		return err
+	}
+	defer zipReader.Close()
+
+	for _, zipFile := range zipReader.File {
+		destFilePath := filepath.Join(dest, zipFile.Name)
+
+		// Zip slip vulnerability check
+		if !strings.HasPrefix(destFilePath, filepath.Clean(dest)+string(os.PathSeparator)) {
+			return fmt.Errorf("%s: illegal file path", destFilePath)
+		}
+
+		if zipFile.FileInfo().IsDir() {
+			err = os.MkdirAll(destFilePath, 0755)
+			if err != nil {
+				return fmt.Errorf("error creating new directory %s", destFilePath)
+			}
+			continue
+		}
+
+		if err := os.MkdirAll(filepath.Dir(zipFilePath), os.ModePerm); err != nil {
+			return err
+		}
+
+		file, fileErr := OpenFile(destFilePath)
+		if fileErr != nil {
+			return fmt.Errorf("unable to open local file '%s' for writing: %w", destFilePath, fileErr)
+		}
+		defer file.Close()
+
+		zippedRc, err := zipFile.Open()
+		if err != nil {
+			return err
+		}
+		defer zippedRc.Close()
+
+		if _, err = io.Copy(file, zippedRc); err != nil {
+			return fmt.Errorf("error writing zip resource to local file '%s': %w", destFilePath, err)
+		}
+
+	}
+	return nil
+}
+
+// Extract a tar/tgz/tar.gz archive file into the provided destination directory.
```

Comment:
This won't work currently for a non-gzipped tar, and we should probably also support gzipped non-tar (single file). I think we need some additional differentiation logic e.g. some of the content types should be unambiguous, just `application/gzip` might need to unzip and then inspect (or peek beginning of stream maybe).

Reply:
Based on my testing, this does actually work with non-gzipped tars even with the gzip reader included. 

For gzip only support, as you say, the content-type `application/gzip` can be used for both `.gz` and `.tar.gz`. Would checking for a lone `.gz` extension then sending it through the gzip reader suffice for this? I feel like gzip only files probably aren't nearly as common as the tar.gz use case.